apps/lib/apps.c

```diff
@@ -629,103 +675,119 @@ EVP_PKEY *load_key(const char *file, int format, int maybe_stdin,
     }
 
  end:
-    BIO_free(key);
-    if (pkey == NULL && desc != NULL) {
+    BIO_free(bio);
+    if (pkey == NULL) {
         BIO_printf(bio_err, "Unable to load %s\n", desc);
         ERR_print_errors(bio_err);
     }
     return pkey;
 }
 
-EVP_PKEY *load_pubkey(const char *file, int format, int maybe_stdin,
+EVP_PKEY *load_pubkey(const char *uri, int format, int maybe_stdin,
                       const char *pass, ENGINE *e, const char *desc)
 {
-    BIO *key = NULL;
+    BIO *bio = NULL;
     EVP_PKEY *pkey = NULL;
     PW_CB_DATA cb_data;
 
+    if (desc == NULL)
+        desc = "public key";
     cb_data.password = pass;
-    cb_data.prompt_info = file;
+    cb_data.prompt_info = uri;
 
-    if (file == NULL && (!maybe_stdin || format == FORMAT_ENGINE)) {
-        BIO_printf(bio_err, "No keyfile specified\n");
-        goto end;
-    }
     if (format == FORMAT_ENGINE) {
         if (e == NULL) {
-            BIO_printf(bio_err, "No engine specified\n");
+            BIO_printf(bio_err, "No engine specified for loading %s\n", desc);
         } else {
 #ifndef OPENSSL_NO_ENGINE
-            pkey = ENGINE_load_public_key(e, file, (UI_METHOD *)get_ui_method(),
+            pkey = ENGINE_load_public_key(e, uri, (UI_METHOD *)get_ui_method(),
                                           &cb_data);
-            if (pkey == NULL && desc != NULL) {
+            if (pkey == NULL) {
                 BIO_printf(bio_err, "Cannot load %s from engine\n", desc);
                 ERR_print_errors(bio_err);
             }
 #else
-            BIO_printf(bio_err, "Engines not supported\n");
+            BIO_printf(bio_err, "Engines not supported for loading %s\n", desc);
 #endif
         }
         goto end;
     }
-    if (file == NULL && maybe_stdin) {
+    if (uri == NULL){
+        if (!maybe_stdin) {
+            BIO_printf(bio_err, "No filename or uri specified for loading %s\n",
+                       desc);
+        }
         unbuffer(stdin);
-        key = dup_bio_in(format);
-    } else {
-        key = bio_open_default(file, 'r', format);
+        bio = dup_bio_in(format);
```

Comment:
Again, `OSSL_STORE_attach`.  That will probably also simplify your code, as you really shouldn't need to care about the format at all, except for ENGINE.

Reply:
Yes, and for the exotic legacy formats FORMAT_MSBLOB and FORMAT_PVK.

Reply:
Why?

Reply:
Oops, these two legacy formats FORMAT_MSBLOB and FORMAT_PVK are relevant for load_key, while for load_pubkey it is only FORMAT_MSBLOB.
Or does OSSL_STORE support these as well?

Reply:
I've verified that the legacy format are not supported by OSSL_STORE,
so in #11755 I kept their support in `apps/lib/apps.c`.